BUILD

```diff
@@ -34,3 +34,11 @@ filegroup(
     ],
     visibility = ["//distro:__pkg__"],
 )
+
+filegroup(
+    name = "bzl_srcs",
+    srcs = glob(["**/*.bzl"]) + [
```

Comment:
If we go with this,  these should all
- not use glob
- restrict visibility to the places that can consume it. That might be broader for pkg.bzl, but certainly locked down for distro.

Can you file issues for this kind of thing first so we can discuss the right solutions first.

Reply:
The use of `glob` pairs well with the release pattern here of creating a `.tar` file with select files. It's intentional that these targets broadly include everything in the package to allow the `package_contents` targets to remain the source of truth.

Based on my experience, there aren't many other solutions. This pattern of filegroups is used in `rules_python`, `rules_docker, `rules_nodejs`, `rules_cc`, `rules_rust` and surely others. If the glob pattern is a large issue for you, that can be changed but I'd warn that it's highly vulnerable to breakages with explicit packages. If a single file is missing, the ability to generate docs is broken and would require a PR and a new release.

Reply:
Sure, but the ability to generate docs is not a release breaker.
The doc generation code is not part of the released tarball.

The bigger problem is that I do not want to make a commitment that the stardoc rules will always work, or be useful.  I have a PR out #474 that starts to move the process towards post-processing the stardoc to generate a unified reference page.
The intention is that we'll add code in the future to generate inter rule links, turn markers into version specific stamps and several other things that we've been thinking about in the overall bazel docs rewrite.

Let's say you could point at @rules_pkg//pkg:bzl_files and get a set of every .bzl file.  I am not about to make a commitment that they can be consumed for any other purpose than using them in rules.  You might get stardoc to produce something, but that won't necessarily be good docs.

Here's an example we already have which shows the problem:  All the common attributes are documented once. I won't accept PRs that expand the text into the rule declarations in the .bzl files..  But, we could go with a path that post processes stardoc output and expands the boilerplate text there.  Either way, the raw stardoc generated text is insufficient.  The promise to users that I want to make is that we delivere high quality docs alongside the distribution. I won't promise that our source tree is arranged in such a way that people can bring their own tools to produce docs.

Reply:
I don't think we're on the same page. I have rules/macros that use `rules_pkg`. I want to generate documentation using `stardoc` so employees of `${WORKPLACE}` have some documentation they can read that explains how to use the rules/macros without having to read the code itself. This works fine with `rules_pkg-0.2.1` but in updating to `rules_pkg-0.5.1`, I can no longer generate my own docs because of a change to some `.bzl` file that includes a private `.bzl` file. For me, it would be ideal to not need to build a `bzl_library` of every individual file `rules_pkg` loads transitively. Instead, I want a single target that has everything I need. That is what this pull-request provides in the form of a `filegroup`. If the `bzl_library` in [@rules_pkg//distro](https://github.com/bazelbuild/rules_pkg/blob/ce8c20ac803fecdb99111a6d7b579e5f51146d1a/distro/BUILD#L108-L124) were made public and contained all the `*.bzl` files, I'd be happy to use that.

Reply:
I get the desire to generate your own docs.
On my side, I don't want to be hamstrung by an expectation that we'll keep the docs stardoc compliant. I've recently moved doc generation out of distro and into doc_build, where I can build a unified docs page to publish.  See https://bazelbuild.github.io/rules_pkg/0.5.0/reference.html for an example.

I am only using stardoc today because that is what is available.  What I would like to be able to do is add annotations that we can expand in post processing the stardoc output. What I can do is make doc_build available (which is equivalent to your alternative of making distro available).  I just have to figure out the implications of not forcing users to take a stardoc dependency. I could just say don't do  `bazel test @rules_pkg//...`, or maybe I can make it optional in the deps.

Reply:
I think there's a massive amount of confusion here and I don't know how to resolve it. I do not think these rules should have any reliance on `stardoc`. Users should be able to `bazel build @rules_pkg//...` from their repo and never be hit with `@io_bazel_stardoc is missing`. The issue is, I use stardoc for my own docs and after version ~2 I'm not able to generate docs anymore because of some change that added a `*.bzl` file that is no longer visible to *my* stardoc target. All I'm asking for is a filegroup (or `bzl_library`) that has a all `*.bzl` files in the distro package. Which I think is a non-invasive change that puts almost no burden on maintainers (only burden being if a new package is added then the relevant target be made and wired up with the top level `*.bzl` glob).

My situation is I have some wrapper rules around `rules_pkg` and want to generate documentation for them. I cannot because certain `*.bzl` files are not visible to me. What can I do to fix this?

Reply:
I'm not a fan of the bzl_srcs recursive thing, but it is the path of least resistance for getting  all the .bzl file in one place.
I can take this PR, but I don't want to switch doc_build over to that, so I will revert that part and try to merge.

Reply:
Actually, I am still looking for somehting else
Can we just make the sources available at the top level.
Ideally, I would drop an dependency on bzl_skylib, Aside from paths (which can be eliminated), the only reason for it is for bzl_library.

Reply:
That's fine. Though I don't see a reason why you wouldn't use the top level `bzl_library`. `stardoc` has no problem with `bzl_library` targets including more sources that what's used. I've found this to be easier to manage but I may misunderstand the intent of `bzl_library`. Happy learn more about it though